### ISSUE TITLE
Metric 정의하기

### ISSUE THREAD

**Issue by eubinecto** [eubinecto/wisdomify#16]
이제 qualitative analysis는 어느정도 마친 것 같다. 
quantitative analysis를 위해서 metric을 정의를 하고  구현을 하자.
가능하면 pytorch-lightning으로 이걸 어떻게 하는지를 생각해보는걸로! 

Yan et al.  논문 참고 - median & at10, at100, at1000 rank.

## To-do's
- [x]  논문에서 어떤 metric을 사용했는지 확인하기.
- [x] 어떻게 구현하면 될지 리서치 해보기.
- [x] `RDMetrics` 구현하기 (median, var, top1, top10, top100)
- [x]  `test_metrics.py` 작성하고, metric이 잘 작동하는지 테스트하기.
- [x] `widomify/main/eval.py`에 `version_0`의 training performance 구하는 로직짜기

## 떠오르는 문제들
- 각 버전별로 데이터는 어떻게 version control을 해야하는가? (현재는 그냥.. `wisdom2def` 하나로 퉁치고 있다).
  - 각 모델의 버전에 맞추어야 한다. 데이터도.

**Comment by eubinecto** [comment]
논문의 result 섹션을 보면, 다음과 같은  metric을 사용한다.

<img width="642" alt="image" src="https://user-images.githubusercontent.com/56193069/124703045-84881d00-df2c-11eb-880b-96d3f7aa5b77.png">

1. median rank:  전체 sample
2. Acc at 1, 10, 100
3. rank variance

일단, 이 셋을 구현하는 것을 첫번째 목표로 삼아야 한다.

**Comment by eubinecto** [comment]
논문의 저자는 이걸 어떻게 구현했나... 참고할 필요가 있다.  다시한번 살펴보자. (design & implementation)

**Comment by eubinecto** [comment]
## `pytorch-lightning` 에서도 비슷한 것을 햘 수 있지 않을까?

**Comment by eubinecto** [comment]
## `torchmetrics` API 톫아보기

먼저, 설치부터:
```bash
pip3 install torchmetrics
```

**Comment by eubinecto** [comment]
## `RDMetrics` 구현하기

`wisdomify/metrics.py`에다가  `RDMetrics` 클래스를 구현하자.

**Comment by eubinecto** [comment]
## `RDMetric` 테스트 하기

`wisdomify/test_metrics.py`에다가, `TestRDMetrics`을 구현해보자.

... 다음과 같이 unittest를 작성했다. 생각보다 간단해서  유닛테스트를 하는 것이 크게 어렵지는 않은 듯.
```python3
class TestRDMetric(unittest.TestCase):

    preds: torch.Tensor  # predictions
    targets: torch.LongTensor  # targets (labels)
    ranks: List[int]

    @classmethod
    def setUpClass(cls) -> None:
        cls.preds = torch.Tensor([[0.6, 0.3, 0.1],
                                  [0.3, 0.1, 0.6],
                                  [0.1, 0.6, 0.3]])  # should be normalised
        cls.targets = torch.LongTensor([1, 1, 1])  # should be a long tensor
        cls.ranks = [1, 2, 0]

    def test_update(self):
        rd_metric = RDMetric()
        rd_metric.update(self.preds, self.targets)
        self.assertEqual(self.ranks, rd_metric.ranks)
        self.assertEqual(1, rd_metric.correct_top1)
        self.assertEqual(3, rd_metric.correct_top10)
        self.assertEqual(3, rd_metric.correct_top100)
        self.assertEqual(self.targets.numel(), rd_metric.total)

    def test_compute(self):
        rd_metric = RDMetric()
        rd_metric.update(self.preds, self.targets)
        median, var, top1, top10, top100 = rd_metric.compute()
        self.assertEqual(np.median(self.ranks), median)
        self.assertEqual(np.var(self.ranks), var)
```


위의 테스트를 작성하면서, `RDMetric`에 문제가 여러군데 있었음을 발견함. 이를 전부 수정했음:
```python3
    def update(self, preds: torch.Tensor, targets: torch.Tensor) -> None:  # noqa
        # so, how do I get the ranks?
        argsorted = torch.argsort(preds, dim=1, descending=True)
        self.ranks = (argsorted == targets).nonzero(as_tuple=True)[1].tolist()  # noqa, see examples/explore_nonzero
        self.correct_top1 = torch.eq(argsorted[:, :1], targets[:1]).sum()  # noqa
        self.correct_top10 = torch.eq(argsorted[:, :10], targets[:10]).sum()  # noqa
        self.correct_top100 = torch.eq(argsorted[:, :100], targets[:100]).sum()  # noqa
        self.total += targets.numel()

    def compute(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
        """
        returns: median, var, top1 acc, top10 acc, top10 acc.
        """
        median = np.median(self.ranks)
        var = np.var(self.ranks)
        top1_acc = self.correct_top1.float() / self.total
        top10_acc = self.correct_top10.float() / self.total
        top100_acc = self.correct_top100.float() / self.total
        return median, var, top1_acc, top10_acc, top100_acc
```
- `argsort`를 할때, descending order로 소팅하는 것을 깜박했었음. 수정함.
- `targets`가 1차원 텐서였음을 깜박하고 슬라이싱을 두 차원 모두에 걸쳐서 했었음. 첫번째 차원만 슬라이싱하는 것으로 수정함.
- `ranks`가  list 오브젝트인데 `torch.median`, `torch.var`로  값을 구하고 있었음. `numpy`함수를 사용하는 것으로 수정.


이제 테스트는 전부 패스함! `RDMetric`은 구현완료!
이제 이걸로 훈련할 때, 훈련 끝나고 나서, 모델을 평가할 때 쓰면 된다!

**Comment by eubinecto** [comment]
## `wisdomify/main/eval.py` 로  `version_0`의 training performance 측정하기
